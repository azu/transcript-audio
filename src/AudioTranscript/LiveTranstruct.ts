export type LiveTranscript = { text: string; currentTime: number };
export type IRTranscript = {
    startIndex: number;
    endIndex: number;
    startTime: number;
    endTime: number;
};
export type createLiveTranscriptResultItem = {
    startIndex: number;
    endIndex: number;
    text: string;
    startTime: number;
    endTime: number;
};
export type createLiveTranscriptResult = {
    text: string;
    items: createLiveTranscriptResultItem[];
};
export const createLiveTranscript = (startTime: number = 0) => {
    let rawTranscripts: LiveTranscript[] = [];
    let internalTranscripts: IRTranscript[] = [];
    return {
        valid(transcript: LiveTranscript) {
            const lastTranscript = rawTranscripts[rawTranscripts.length - 1];
            if (!lastTranscript) {
                return true;
            }
            // if text length is lower than previous, it is invalid
            return transcript.text.length > lastTranscript.text.length;
        },
        add(transcript: LiveTranscript) {
            if (!this.valid(transcript)) {
                return;
            }
            rawTranscripts.push(transcript);
            const lastIR = internalTranscripts[internalTranscripts.length - 1];
            if (lastIR) {
                internalTranscripts.push({
                    startIndex: lastIR.endIndex,
                    endIndex: transcript.text.length,
                    startTime: lastIR.endTime + 1,
                    endTime: transcript.currentTime
                });
            } else {
                // first IR
                internalTranscripts.push({
                    startIndex: 0,
                    endIndex: transcript.text.length,
                    startTime: startTime,
                    endTime: transcript.currentTime
                });
            }
        },
        get(): createLiveTranscriptResult {
            const lastTranscript = rawTranscripts[rawTranscripts.length - 1];
            if (!lastTranscript) {
                return {
                    text: "",
                    items: []
                };
            }
            return {
                text: lastTranscript.text,
                items: internalTranscripts.map((ir) => {
                    return {
                        ...ir,
                        text: lastTranscript.text.slice(ir.startIndex, ir.endIndex)
                    };
                })
            };
        },
        clear() {
            rawTranscripts = [];
            internalTranscripts = [];
        },
        setStartTime(newStartTime: number) {
            startTime = newStartTime;
        }
    };
};
