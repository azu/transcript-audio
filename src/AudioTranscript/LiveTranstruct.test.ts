import { createLiveTranscript } from "./LiveTranstruct";
import fixture from "./LiveTranstruct.test.json";

describe("LiveTranscript", function () {
    it("example", () => {
        const tr = createLiveTranscript();
        fixture.forEach((item) => {
            tr.add(item);
        });
        expect(tr.get()).toMatchSnapshot();
    });
});
