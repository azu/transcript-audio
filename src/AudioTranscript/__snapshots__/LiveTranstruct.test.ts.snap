// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LiveTranscript example 1`] = `
Object {
  "items": Array [
    Object {
      "endIndex": 7,
      "endTime": 6.133043,
      "startIndex": 0,
      "startTime": 0,
      "text": "アイポッドキャ",
    },
    Object {
      "endIndex": 8,
      "endTime": 6.283739,
      "startIndex": 7,
      "startTime": 7.133043,
      "text": "ス",
    },
    Object {
      "endIndex": 9,
      "endTime": 6.476441,
      "startIndex": 8,
      "startTime": 7.283739,
      "text": "ト",
    },
    Object {
      "endIndex": 11,
      "endTime": 6.785899,
      "startIndex": 9,
      "startTime": 7.476441,
      "text": "始め",
    },
    Object {
      "endIndex": 13,
      "endTime": 7.476428,
      "startIndex": 11,
      "startTime": 7.785899,
      "text": "まー",
    },
    Object {
      "endIndex": 17,
      "endTime": 12.282523,
      "startIndex": 13,
      "startTime": 8.476428,
      "text": "ちゃんと",
    },
    Object {
      "endIndex": 19,
      "endTime": 12.583895,
      "startIndex": 17,
      "startTime": 13.282523,
      "text": "言え",
    },
    Object {
      "endIndex": 20,
      "endTime": 12.6856,
      "startIndex": 19,
      "startTime": 13.583895,
      "text": "て",
    },
    Object {
      "endIndex": 23,
      "endTime": 12.803959,
      "startIndex": 20,
      "startTime": 13.6856,
      "text": "言って",
    },
    Object {
      "endIndex": 24,
      "endTime": 12.974257,
      "startIndex": 23,
      "startTime": 13.803959,
      "text": "く",
    },
    Object {
      "endIndex": 25,
      "endTime": 12.97749,
      "startIndex": 24,
      "startTime": 13.974257,
      "text": "だ",
    },
    Object {
      "endIndex": 26,
      "endTime": 12.980978,
      "startIndex": 25,
      "startTime": 13.97749,
      "text": "さ",
    },
    Object {
      "endIndex": 27,
      "endTime": 13.078165,
      "startIndex": 26,
      "startTime": 13.980978,
      "text": "っ",
    },
    Object {
      "endIndex": 28,
      "endTime": 13.274864,
      "startIndex": 27,
      "startTime": 14.078165,
      "text": "て",
    },
    Object {
      "endIndex": 30,
      "endTime": 13.278796,
      "startIndex": 28,
      "startTime": 14.274864,
      "text": "あり",
    },
    Object {
      "endIndex": 33,
      "endTime": 13.473681,
      "startIndex": 30,
      "startTime": 14.278796,
      "text": "がとう",
    },
    Object {
      "endIndex": 36,
      "endTime": 13.676034,
      "startIndex": 33,
      "startTime": 14.473681,
      "text": "ござい",
    },
    Object {
      "endIndex": 38,
      "endTime": 13.782331,
      "startIndex": 36,
      "startTime": 14.676034,
      "text": "ます",
    },
    Object {
      "endIndex": 40,
      "endTime": 16.183386,
      "startIndex": 38,
      "startTime": 14.782331,
      "text": "今日",
    },
    Object {
      "endIndex": 41,
      "endTime": 16.376554,
      "startIndex": 40,
      "startTime": 17.183386,
      "text": "は",
    },
    Object {
      "endIndex": 43,
      "endTime": 17.128466,
      "startIndex": 41,
      "startTime": 17.376554,
      "text": "大手",
    },
    Object {
      "endIndex": 44,
      "endTime": 17.190767,
      "startIndex": 43,
      "startTime": 18.128466,
      "text": "歳",
    },
    Object {
      "endIndex": 45,
      "endTime": 17.290705,
      "startIndex": 44,
      "startTime": 18.190767,
      "text": "の",
    },
    Object {
      "endIndex": 46,
      "endTime": 17.342465,
      "startIndex": 45,
      "startTime": 18.290705,
      "text": "フ",
    },
    Object {
      "endIndex": 47,
      "endTime": 17.661718,
      "startIndex": 46,
      "startTime": 18.342465,
      "text": "ロ",
    },
    Object {
      "endIndex": 49,
      "endTime": 17.679214,
      "startIndex": 47,
      "startTime": 18.661718,
      "text": "ント",
    },
    Object {
      "endIndex": 52,
      "endTime": 17.975314,
      "startIndex": 49,
      "startTime": 18.679214,
      "text": "エンド",
    },
    Object {
      "endIndex": 53,
      "endTime": 18.280665,
      "startIndex": 52,
      "startTime": 18.975314,
      "text": "エ",
    },
    Object {
      "endIndex": 57,
      "endTime": 18.380556,
      "startIndex": 53,
      "startTime": 19.280665,
      "text": "ンジニア",
    },
    Object {
      "endIndex": 58,
      "endTime": 18.681289,
      "startIndex": 57,
      "startTime": 19.380556,
      "text": "の",
    },
    Object {
      "endIndex": 59,
      "endTime": 18.885004,
      "startIndex": 58,
      "startTime": 19.681289,
      "text": "ト",
    },
    Object {
      "endIndex": 61,
      "endTime": 19.223465,
      "startIndex": 59,
      "startTime": 19.885004,
      "text": "ムさ",
    },
    Object {
      "endIndex": 62,
      "endTime": 19.378397,
      "startIndex": 61,
      "startTime": 20.223465,
      "text": "ん",
    },
    Object {
      "endIndex": 63,
      "endTime": 19.993081,
      "startIndex": 62,
      "startTime": 20.378397,
      "text": "に",
    },
    Object {
      "endIndex": 64,
      "endTime": 20.177386,
      "startIndex": 63,
      "startTime": 20.993081,
      "text": "来",
    },
    Object {
      "endIndex": 65,
      "endTime": 20.18312,
      "startIndex": 64,
      "startTime": 21.177386,
      "text": "て",
    },
    Object {
      "endIndex": 67,
      "endTime": 20.476157,
      "startIndex": 65,
      "startTime": 21.18312,
      "text": "いた",
    },
    Object {
      "endIndex": 69,
      "endTime": 20.579006,
      "startIndex": 67,
      "startTime": 21.476157,
      "text": "だい",
    },
    Object {
      "endIndex": 70,
      "endTime": 20.68524,
      "startIndex": 69,
      "startTime": 21.579006,
      "text": "て",
    },
    Object {
      "endIndex": 71,
      "endTime": 20.875386,
      "startIndex": 70,
      "startTime": 21.68524,
      "text": "ま",
    },
    Object {
      "endIndex": 72,
      "endTime": 21.075345,
      "startIndex": 71,
      "startTime": 21.875386,
      "text": "す",
    },
    Object {
      "endIndex": 74,
      "endTime": 21.224122,
      "startIndex": 72,
      "startTime": 22.075345,
      "text": "無理",
    },
    Object {
      "endIndex": 75,
      "endTime": 21.312124,
      "startIndex": 74,
      "startTime": 22.224122,
      "text": "や",
    },
    Object {
      "endIndex": 77,
      "endTime": 21.409081,
      "startIndex": 75,
      "startTime": 22.312124,
      "text": "さん",
    },
    Object {
      "endIndex": 78,
      "endTime": 21.594762,
      "startIndex": 77,
      "startTime": 22.409081,
      "text": "よ",
    },
    Object {
      "endIndex": 80,
      "endTime": 21.675787,
      "startIndex": 78,
      "startTime": 22.594762,
      "text": "ろし",
    },
    Object {
      "endIndex": 81,
      "endTime": 21.778798,
      "startIndex": 80,
      "startTime": 22.675787,
      "text": "く",
    },
    Object {
      "endIndex": 82,
      "endTime": 21.780544,
      "startIndex": 81,
      "startTime": 22.778798,
      "text": "お",
    },
    Object {
      "endIndex": 83,
      "endTime": 21.976435,
      "startIndex": 82,
      "startTime": 22.780544,
      "text": "願",
    },
    Object {
      "endIndex": 84,
      "endTime": 21.979989,
      "startIndex": 83,
      "startTime": 22.976435,
      "text": "い",
    },
    Object {
      "endIndex": 87,
      "endTime": 22.28057,
      "startIndex": 84,
      "startTime": 22.979989,
      "text": "します",
    },
    Object {
      "endIndex": 92,
      "endTime": 23.813415,
      "startIndex": 87,
      "startTime": 23.28057,
      "text": "トムですい",
    },
    Object {
      "endIndex": 95,
      "endTime": 24.268696,
      "startIndex": 92,
      "startTime": 24.813415,
      "text": "ません",
    },
    Object {
      "endIndex": 96,
      "endTime": 24.304082,
      "startIndex": 95,
      "startTime": 25.268696,
      "text": "よ",
    },
    Object {
      "endIndex": 99,
      "endTime": 24.48505,
      "startIndex": 96,
      "startTime": 25.304082,
      "text": "ろしく",
    },
    Object {
      "endIndex": 100,
      "endTime": 24.576066,
      "startIndex": 99,
      "startTime": 25.48505,
      "text": "お",
    },
    Object {
      "endIndex": 102,
      "endTime": 24.579241,
      "startIndex": 100,
      "startTime": 25.576066,
      "text": "願い",
    },
    Object {
      "endIndex": 103,
      "endTime": 24.778571,
      "startIndex": 102,
      "startTime": 25.579241,
      "text": "し",
    },
    Object {
      "endIndex": 105,
      "endTime": 24.976454,
      "startIndex": 103,
      "startTime": 25.778571,
      "text": "ます",
    },
    Object {
      "endIndex": 107,
      "endTime": 26.175905,
      "startIndex": 105,
      "startTime": 25.976454,
      "text": "よろ",
    },
    Object {
      "endIndex": 109,
      "endTime": 26.280995,
      "startIndex": 107,
      "startTime": 27.175905,
      "text": "しく",
    },
    Object {
      "endIndex": 110,
      "endTime": 26.37938,
      "startIndex": 109,
      "startTime": 27.280995,
      "text": "お",
    },
    Object {
      "endIndex": 112,
      "endTime": 26.47865,
      "startIndex": 110,
      "startTime": 27.37938,
      "text": "願い",
    },
    Object {
      "endIndex": 113,
      "endTime": 26.57651,
      "startIndex": 112,
      "startTime": 27.47865,
      "text": "し",
    },
    Object {
      "endIndex": 115,
      "endTime": 26.581327,
      "startIndex": 113,
      "startTime": 27.57651,
      "text": "ます",
    },
    Object {
      "endIndex": 119,
      "endTime": 27.779314,
      "startIndex": 115,
      "startTime": 27.581327,
      "text": "ちょっと",
    },
    Object {
      "endIndex": 120,
      "endTime": 28.084809,
      "startIndex": 119,
      "startTime": 28.779314,
      "text": "最",
    },
    Object {
      "endIndex": 121,
      "endTime": 28.182727,
      "startIndex": 120,
      "startTime": 29.084809,
      "text": "初",
    },
    Object {
      "endIndex": 122,
      "endTime": 28.379515,
      "startIndex": 121,
      "startTime": 29.182727,
      "text": "に",
    },
    Object {
      "endIndex": 123,
      "endTime": 28.990079,
      "startIndex": 122,
      "startTime": 29.379515,
      "text": "簡",
    },
    Object {
      "endIndex": 124,
      "endTime": 29.080219,
      "startIndex": 123,
      "startTime": 29.990079,
      "text": "単",
    },
    Object {
      "endIndex": 125,
      "endTime": 29.174921,
      "startIndex": 124,
      "startTime": 30.080219,
      "text": "に",
    },
    Object {
      "endIndex": 127,
      "endTime": 29.496413,
      "startIndex": 125,
      "startTime": 30.174921,
      "text": "自己",
    },
    Object {
      "endIndex": 128,
      "endTime": 29.579136,
      "startIndex": 127,
      "startTime": 30.496413,
      "text": "紹",
    },
    Object {
      "endIndex": 129,
      "endTime": 29.683294,
      "startIndex": 128,
      "startTime": 30.579136,
      "text": "介",
    },
    Object {
      "endIndex": 130,
      "endTime": 29.981896,
      "startIndex": 129,
      "startTime": 30.683294,
      "text": "を",
    },
    Object {
      "endIndex": 132,
      "endTime": 30.383557,
      "startIndex": 130,
      "startTime": 30.981896,
      "text": "して",
    },
    Object {
      "endIndex": 133,
      "endTime": 30.485163,
      "startIndex": 132,
      "startTime": 31.383557,
      "text": "も",
    },
    Object {
      "endIndex": 135,
      "endTime": 30.57866,
      "startIndex": 133,
      "startTime": 31.485163,
      "text": "らっ",
    },
    Object {
      "endIndex": 136,
      "endTime": 30.583586,
      "startIndex": 135,
      "startTime": 31.57866,
      "text": "て",
    },
    Object {
      "endIndex": 137,
      "endTime": 30.68068,
      "startIndex": 136,
      "startTime": 31.583586,
      "text": "も",
    },
    Object {
      "endIndex": 139,
      "endTime": 30.992266,
      "startIndex": 137,
      "startTime": 31.68068,
      "text": "良い",
    },
    Object {
      "endIndex": 140,
      "endTime": 31.081252,
      "startIndex": 139,
      "startTime": 31.992266,
      "text": "で",
    },
    Object {
      "endIndex": 143,
      "endTime": 31.182308,
      "startIndex": 140,
      "startTime": 32.081252,
      "text": "しょう",
    },
    Object {
      "endIndex": 144,
      "endTime": 31.287907,
      "startIndex": 143,
      "startTime": 32.182308,
      "text": "か",
    },
    Object {
      "endIndex": 146,
      "endTime": 32.383505,
      "startIndex": 144,
      "startTime": 32.287907000000004,
      "text": "はい",
    },
    Object {
      "endIndex": 148,
      "endTime": 33.697581,
      "startIndex": 146,
      "startTime": 33.383505,
      "text": "トム",
    },
    Object {
      "endIndex": 150,
      "endTime": 33.991791,
      "startIndex": 148,
      "startTime": 34.697581,
      "text": "と言",
    },
    Object {
      "endIndex": 152,
      "endTime": 34.578934,
      "startIndex": 150,
      "startTime": 34.991791,
      "text": "いま",
    },
    Object {
      "endIndex": 153,
      "endTime": 34.583443,
      "startIndex": 152,
      "startTime": 35.578934,
      "text": "す",
    },
    Object {
      "endIndex": 154,
      "endTime": 35.482422,
      "startIndex": 153,
      "startTime": 35.583443,
      "text": "ね",
    },
    Object {
      "endIndex": 156,
      "endTime": 35.688,
      "startIndex": 154,
      "startTime": 36.482422,
      "text": "もと",
    },
    Object {
      "endIndex": 157,
      "endTime": 35.8848,
      "startIndex": 156,
      "startTime": 36.688,
      "text": "も",
    },
    Object {
      "endIndex": 158,
      "endTime": 35.890652,
      "startIndex": 157,
      "startTime": 36.8848,
      "text": "と",
    },
    Object {
      "endIndex": 159,
      "endTime": 36.181759,
      "startIndex": 158,
      "startTime": 36.890652,
      "text": "は",
    },
    Object {
      "endIndex": 163,
      "endTime": 36.694072,
      "startIndex": 159,
      "startTime": 37.181759,
      "text": " Web",
    },
    Object {
      "endIndex": 164,
      "endTime": 37.286722,
      "startIndex": 163,
      "startTime": 37.694072,
      "text": " ",
    },
    Object {
      "endIndex": 165,
      "endTime": 37.412237,
      "startIndex": 164,
      "startTime": 38.286722,
      "text": "の",
    },
    Object {
      "endIndex": 168,
      "endTime": 38.383505,
      "startIndex": 165,
      "startTime": 38.412237,
      "text": "住宅で",
    },
    Object {
      "endIndex": 178,
      "endTime": 40.28692,
      "startIndex": 168,
      "startTime": 39.383505,
      "text": "色々開発をやっていて",
    },
    Object {
      "endIndex": 194,
      "endTime": 43.593247,
      "startIndex": 178,
      "startTime": 41.28692,
      "text": "色々開発をやっていたんですけども",
    },
    Object {
      "endIndex": 214,
      "endTime": 47.487564,
      "startIndex": 194,
      "startTime": 44.593247,
      "text": "えーと社長の家からさんとは古い知り合いで",
    },
  ],
  "text": "アイポッドキャスト始めまーちゃんと言えて言ってくださってありがとうございます今日は大手歳のフロントエンドエンジニアのトムさんに来ていただいてます無理やさんよろしくお願いしますトムですいませんよろしくお願いしますよろしくお願いしますちょっと最初に簡単に自己紹介をしてもらっても良いでしょうかはいトムと言いますねもともとは Web の住宅で色々開発をやっていて色々開発をやっていたんですけどもえーと社長の家からさんとは古い知り合いで",
}
`;
